File: android/app/src/main/java/com/modi/connect/links/usb/UsbLink.kt
/*
 * MoDi Connect - Cross-device interconnection protocol
 * Copyright (C) 2026 Silvite
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.modi.connect.links.usb

import android.content.Context
import android.content.Intent
import android.media.projection.MediaProjection
import com.modi.connect.ConnectionState
import com.modi.connect.ConnectionStateManager
import com.modi.connect.StreamingService
import com.modi.connect.audio.AudioPipeline
import com.modi.protocol.PacketHeaderCodec
import com.modi.connect.core.adapters.UsbTransport
import com.modi.protocol.Packet
import com.modi.connect.links.ILink
import com.modi.connect.links.LinkState
import com.modi.connect.links.LinkManager
import com.modi.connect.links.LinkParams
import com.modi.protocol.LinkType
import com.modi.protocol.PacketType
import com.modi.connect.session.DisconnectReason
import com.modi.connect.session.SessionControlMessage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.util.UUID

/**
 * UsbLink — USB 链路（Android 端，主动发起方）
 *
 * 职责：用户点击"USB 直连" → TCP 连接 localhost:12348 → 发 HELLO → 推流。
 * 与蓝牙链路对称：手机主动发起连接和握手，Windows 常驻等待。
 *
 * 握手方向：Android 发 HELLO(token+route) → Windows 校验 → 回 HELLO_ACK(route)（与蓝牙一致）
 * 数据通路：AudioPipeline → EncodeSender → UsbTransport.sendBlocking()
 *
 * 依赖：USB 链路专属，与 LAN/P2P/蓝牙完全解耦。
 * 前置条件：USB 线已连接 + USB 调试已开启 + Windows 端已自动建立 adb forward。
 */
class UsbLink(
    private val context: Context,
    private val pipe: AudioPipeline,
    private val stateManager: ConnectionStateManager
) : ILink {

    companion object {
        private const val TAG = "UsbLink"
    }

    // ── 子模块 ──
    private val connectMutex = Mutex()
    private var usbTransport: UsbTransport? = null

    // ── ILink 状态 ──
    @Volatile override var isStreaming = false
        private set
    override val state: LinkState
        get() = if (isStreaming) LinkState.STREAMING else LinkState.IDLE
    override var onStatusChanged: ((String) -> Unit)? = null
    override var onStreamingChanged: ((Boolean) -> Unit)? = null
    @Volatile override var sessionId: UUID? = null
        private set

    @Volatile var currentRoute: Int = 0
        private set

    // ── ILink 实现 ──

    /**
     * 连接 USB 链路（用户点击“USB 直连”触发）
     * 流程：启动 TCP Server → 等待 Windows 连接 → 发 HELLO → 等 ACK → 推流
     */
    override suspend fun connect(params: LinkParams): Boolean = connectMutex.withLock {
        currentRoute = params.route
        onStatusChanged?.invoke("USB：启动监听，请确认 USB 已连接且电脑端已就绪...")
        stateManager.update(ConnectionState.CONNECTING)
    
        // 1. 启动 TCP Server（等待 Windows 通过 adb forward 连入）
        val transport = UsbTransport()
        transport.startListening()
        usbTransport = transport
    
        val connected = try {
            // 等待 ServerSocket 绑定完成
            delay(200)

            // 2. 等待 Windows 连接
            onStatusChanged?.invoke("USB：等待电脑连接...")
            transport.waitForConnection(timeoutMs = 0)
        } catch (cancelled: CancellationException) {
            transport.stopListening()
            usbTransport = null
            sessionId = null
            throw cancelled
        }
        if (!connected) {
            onStatusChanged?.invoke("USB：等待连接超时")
            stateManager.update(ConnectionState.ERROR)
            transport.stopListening()
            usbTransport = null
            return@withLock false
        }
        onStatusChanged?.invoke("USB：电脑已连接，握手中...")

        // 2. 主动发 HELLO → 等待 ACK（与蓝牙 BtHandshakeClient 对称）
        val handshake = withContext(Dispatchers.IO) {
            UsbHandshakeClient.sendHelloAndWaitForAck(transport, params.route, UUID.randomUUID())
        }
        if (handshake == null) {
            onStatusChanged?.invoke("USB：握手失败（电脑未响应）")
            stateManager.update(ConnectionState.ERROR)
            disconnect()
            return@withLock false
        }

        stateManager.update(ConnectionState.CONNECTED)
        sessionId = handshake.sessionId
        onStatusChanged?.invoke("USB：握手成功 ✓ 准备推流")

        // 3. 启动推流（注入 UsbTransport）
        val capMode = LinkManager.routeToCapture(params.route)
        pipe.onFirstFrame = { stateManager.update(ConnectionState.STREAMING) }
        val ok = withContext(Dispatchers.IO) {
            pipe.currentLinkType = LinkType.USB
            pipe.startStreamingWithTransport(transport, capMode, params.proj, context)
        }

        if (ok) {
            isStreaming = true
            onStatusChanged?.invoke("USB 推流中：路线${params.route + 1}")
            onStreamingChanged?.invoke(true)
            context.startForegroundService(Intent(context, StreamingService::class.java))
        } else {
            onStatusChanged?.invoke("USB：启动推流失败")
            stateManager.update(ConnectionState.ERROR)
            disconnect()
        }
        ok
    }

    /** 推流中热切路线：切换采集模式 + 发送 ROUTE 包通知 Windows 切换 AudioRouter */
    override suspend fun sendRouteUpdate(route: Int, proj: MediaProjection?): Boolean {
        if (!isStreaming) { currentRoute = route; return true }
        currentRoute = route

        val capMode = LinkManager.routeToCapture(route)
        val ok = withContext(Dispatchers.IO) { pipe.switchMode(capMode, proj, context) }
        if (!ok) { onStatusChanged?.invoke("需先授权系统音频"); return false }

        // 发送 ROUTE 包到 Windows
        val transport = usbTransport ?: return false
        withContext(Dispatchers.IO) {
            val protocol = PacketHeaderCodec()
            val payload = byteArrayOf(route.toByte())
            val packet = Packet(PacketType.ROUTE, LinkType.USB, 0u, payload)
            transport.sendBlocking(protocol.encode(packet))
        }
        return true
    }

    override suspend fun sendDisconnectRequest(targetLink: Byte, reason: DisconnectReason): Boolean {
        val currentId = sessionId ?: return false
        val transport = usbTransport?.takeIf { it.isConnected } ?: return false
        val message = SessionControlMessage.request(currentId, LinkType.USB, targetLink, reason)
        return withContext(Dispatchers.IO) {
            val encoded = PacketHeaderCodec().encode(message.toPacket())
            transport.sendBlocking(encoded)
            transport.isConnected
        }
    }

    /** 断开 USB 链路：停止推流 + 关闭 TCP Server + 状态回退 */
    override suspend fun disconnect() {
        context.stopService(Intent(context, StreamingService::class.java))
        pipe.stopStreaming()

        usbTransport?.stopListening()
        usbTransport = null
        sessionId = null

        isStreaming = false
        stateManager.update(ConnectionState.DISCONNECTED)
        onStatusChanged?.invoke("USB：已停止")
        onStreamingChanged?.invoke(false)
    }

}

File: windows/MoDi.Desktop/Links/Usb/UsbLink.cs
/*
 * MoDi Connect - Cross-device interconnection protocol
 * Copyright (C) 2026 Silvite
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
using System;
using System.Threading;
using System.Threading.Tasks;
using MoDi.Core;
using MoDi.Protocol;
using MoDi.Core.Adapters;
using MoDi.Core.Factory;
using MoDi.Core.Infrastructure;
using MoDi.Desktop.Core.Session;
using MoDi.Desktop.Diagnostics;

namespace MoDi.Desktop.Links;

/// <summary>
/// UsbLink — USB 链路（Windows 端，常驻监听）
///
/// 职责：后台循环检测 USB 设备 → 自动 adb forward → TCP 连接手机 → 被动收 HELLO 回 ACK → AudioEngine 播放。
/// 与蓝牙链路对称：Windows 常驻等待，手机主动发起连接和握手。
///
/// 握手方向：Android 发 HELLO(token+route) → Windows 校验 → 回 HELLO_ACK(route)（与 LAN/蓝牙一致）
/// 数据通路：UsbTransport.PacketReceived → AudioEngine（Opus 解码 → 播放）
///
/// 依赖：USB 链路专属，与 LAN/P2P/蓝牙完全解耦。
/// 前置条件：USB 线连接 + 手机开启 USB 调试 + 系统 PATH 有 adb。
/// </summary>
public sealed class UsbLink : ILink
{
    private const string Tag = "UsbLink";

    // ── 链路常量 ──
    private const int DetectIntervalMs = 5_000;  // USB 设备检测轮询间隔

    // ── 核心模块 ──
    private UsbTransport? _transport;
    private AudioEngine? _engine;

    /// <summary>当前活跃的 AudioEngine（会话期间非 null，LinkManager 用于 Volume 控制）</summary>
    public AudioEngine? ActiveEngine => _engine;
    private readonly ConnectionStateManager _stateManager;
    private CancellationTokenSource? _cts;
    private Task? _listenLoop;
    private volatile bool _started;

    // ── 事件（LinkManager / UI 订阅） ──
    public Action<string>? OnStatusChanged;
    public Action<bool>? OnActiveChanged;
    public Action<int>? OnRouteChanged;
    /// <summary>USB 会话开始（手机连接+握手成功），LinkManager 用于暂停 LAN 引擎</summary>
    public Action<Guid>? OnSessionStarted;
    /// <summary>USB 会话结束（手机断开），LinkManager 用于恢复 LAN 引擎</summary>
    public Action<Guid>? OnSessionEnded;
    public Func<SessionControlMessage, byte, SessionControlMessage>? OnDisconnectRequest;

    public LinkState State { get; private set; } = LinkState.Idle;
    public bool IsActive => _started;

    public UsbLink(ConnectionStateManager stateManager)
    {
        _stateManager = stateManager;
    }

    // ── ILink 实现 ──

    /// <summary>启动 USB 链路（常驻监听，与 LAN/蓝牙一起开机启动）</summary>
    public Task<bool> ConnectAsync()
    {
        if (_started) return Task.FromResult(true);
        _started = true;
        State = LinkState.Listening;

        _cts = new CancellationTokenSource();
        var ct = _cts.Token;

        // 启动常驻监听循环（后台）
        _listenLoop = Task.Run(() => ListenLoopAsync(ct));

        OnStatusChanged?.Invoke("USB：就绪，等待手机连接");
        OnActiveChanged?.Invoke(true);
        Log.I(Tag, "USB link started (resident)");
        return Task.FromResult(true);
    }

    /// <summary>停止 USB 链路（关闭监听 + 断开当前连接）</summary>
    public async Task DisconnectAsync()
    {
        if (!_started) return;
        _started = false;
        State = LinkState.Idle;

        var owner = _cts;
        var ownedCancellation = owner?.Token ?? CancellationToken.None;
        owner?.Cancel();

        if (_listenLoop != null)
        {
            await TeardownObserver.AwaitAsync(
                _listenLoop,
                ownedCancellation,
                "USB_LISTEN_LOOP_STOPPED").ConfigureAwait(false);
            _listenLoop = null;
        }

        owner?.Dispose();
        if (ReferenceEquals(_cts, owner))
            _cts = null;

        await CleanupSessionAsync();
        await UsbDeviceHelper.RemoveForwardAsync();

        OnStatusChanged?.Invoke("USB：已停止");
        OnActiveChanged?.Invoke(false);
    }

    // ── 常驻监听循环（核心状态机） ──
    // 流程：轮询检测 USB → adb forward → TCP Server 等待连接 → 被动握手 → 引擎播放 → 等待断开 → 循环

    private async Task ListenLoopAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                // 1. 轮询检测 USB 设备
                OnStatusChanged?.Invoke("USB：等待手机 USB 连接...");
                var hasDevice = await WaitForUsbDeviceAsync(ct);
                if (!hasDevice) continue;

                // 2. 建立 adb forward
                OnStatusChanged?.Invoke("USB：检测到设备，建立隧道...");
                var forwardOk = await UsbDeviceHelper.SetupForwardAsync(ct);
                if (!forwardOk)
                {
                    OnStatusChanged?.Invoke("USB：adb forward 失败，重试...");
                    await Task.Delay(DetectIntervalMs, ct);
                    continue;
                }

                // 3. TCP 连接手机（通过 adb forward 隧道）
                OnStatusChanged?.Invoke("USB：隧道就绪，等待手机连接...");
                _transport = new UsbTransport();
                await _transport.ConnectAsync(ct);

                // 4. 被动握手：等待手机 HELLO → 校验 → 回 ACK
                OnStatusChanged?.Invoke("USB：手机已连接，等待握手...");
                var handshake = await UsbPassiveHandshake.WaitForHelloAsync(_transport, ct);
                if (!handshake.HasValue)
                {
                    OnStatusChanged?.Invoke("USB：握手失败，重新等待...");
                    await CleanupSessionAsync();
                    continue;
                }

                // 5. 创建 AudioEngine → 播放
                _stateManager.BeginConnecting();
                var sessionId = handshake.Value.SessionId;
                OnSessionStarted?.Invoke(sessionId);
                StartAudioEngine(handshake.Value.Route);
                OnRouteChanged?.Invoke(handshake.Value.Route);
                OnStatusChanged?.Invoke($"USB：推流中 ✓ 路线{handshake.Value.Route + 1}");

                // 6. 监听 ROUTE 热切包
                _transport.PacketReceived += OnSessionPacket;


                // 7. 等待连接断开
                await WaitForDisconnectAsync(ct);

                // 8. 断开 → 清理 → 重新等待
                _transport.PacketReceived -= OnSessionPacket;
                OnSessionEnded?.Invoke(sessionId);
                OnStatusChanged?.Invoke("USB：手机断开，重新等待...");
                CleanupAudioEngine();
                await _transport.DisconnectAsync();
                _transport.Dispose();
                _transport = null;
            }
            catch (OperationCanceledException) { break; }
            catch (Exception ex)
            {
                Log.E(Tag, $"ListenLoop error: {ex.Message}");
                if (!ct.IsCancellationRequested)
                    await Task.Delay(DetectIntervalMs, ct);
            }
        }
    }

    /// <summary>轮询等待 USB 设备连接（每 5 秒检测一次 adb devices）</summary>
    private async Task<bool> WaitForUsbDeviceAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            if (await UsbDeviceHelper.DetectDeviceAsync(ct))
                return true;
            await Task.Delay(DetectIntervalMs, ct);
        }
        return false;
    }

    // ── ROUTE 热切（命名方法，确保可取消订阅） ──

    private void OnSessionPacket(ReadOnlyMemory<byte> data)
    {
        var protocol = PlatformFactory.CreateProtocol();
        var decoded = protocol.Decode(data.Span);
        if (!decoded.HasValue) return;
        if (decoded.Value.Type == PacketType.Route)
        {
            UsbPassiveHandshake.HandleRoutePacket(data, _engine, OnStatusChanged, OnRouteChanged);
            return;
        }
        if (decoded.Value.Type != PacketType.Data ||
            !SessionControlMessage.TryDecode(decoded.Value.Payload, out var request) ||
            request.Action != SessionControlAction.DisconnectRequest ||
            OnDisconnectRequest is not { } handleDisconnect)
            return;

        var ack = handleDisconnect(request, LinkType.Usb);
        _ = AcknowledgeThenCloseAsync(protocol, ack);
    }

    private async Task AcknowledgeThenCloseAsync(IPacketProtocol protocol, SessionControlMessage ack)
    {
        var transport = _transport;
        if (transport == null) return;
        await transport.SendAsync(protocol.Encode(ack.ToPacket()));
        if (ack.Result == DisconnectResult.Accepted)
            await transport.DisconnectAsync();
    }

    internal void StopCurrentSession(bool closeTransport)
    {
        CleanupAudioEngine();
        if (closeTransport && _transport != null)
            _ = _transport.DisconnectAsync();
    }

    // ── AudioEngine 管理（USB 专属引擎，不复用 LAN 引擎） ──

    private void StartAudioEngine(int route)
    {
        var speaker = PlatformFactory.CreateRenderer(useCable: false);
        var cable = PlatformFactory.CreateRenderer(useCable: true);
        _engine = new AudioEngine(_transport, speaker, cable);
        _engine.Router.SetMode(UsbPassiveHandshake.RouteToMode(route));

        _engine.OnFirstFrameDecoded += () => _stateManager.Update(ConnectionState.Streaming);
        _engine.Start();
        _stateManager.Update(ConnectionState.Connected);
    }

    private void CleanupAudioEngine()
    {
        _engine?.Stop();
        _engine?.Dispose();
        _engine = null;
    }

    private async Task WaitForDisconnectAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested && _transport?.IsConnected == true)
            await Task.Delay(500, ct);
    }

    private async Task CleanupSessionAsync()
    {
        CleanupAudioEngine();
        if (_transport != null)
        {
            await _transport.DisconnectAsync();
            _transport.Dispose();
            _transport = null;
        }
    }

    public void Dispose()
    {
        _cts?.Cancel();
        _cts?.Dispose();
        _engine?.Dispose();
        _transport?.Dispose();
    }
}
